src/org/mozilla/javascript/NativeJSON.java

```diff
@@ -288,23 +288,42 @@ private static Object str(Object key, Scriptable holder, StringifyState state) {
         Object value = null;
         Object unwrappedJavaValue = null;
 
+        //
+        String keyString = null;
+        int keyInt = 0;
         if (key instanceof String) {
-            value = getProperty(holder, (String) key);
+            keyString = (String) key;
+            value = getProperty(holder, keyString);
         } else {
-            value = getProperty(holder, ((Number) key).intValue());
+            keyInt = ((Number) key).intValue();
+            value = getProperty(holder, keyInt);
         }
 
         if (value instanceof Scriptable && hasProperty((Scriptable) value, "toJSON")) {
             Object toJSON = getProperty((Scriptable) value, "toJSON");
             if (toJSON instanceof Callable) {
-                value = callMethod(state.cx, (Scriptable) value, "toJSON", new Object[] {key});
+                value =
+                        callMethod(
+                                state.cx,
+                                (Scriptable) value,
+                                "toJSON",
+                                new Object[] {
+                                    keyString == null ? Integer.toString(keyInt) : keyString
+                                });
             }
         } else if (value instanceof BigInteger) {
```

Comment:
Although you didnt change this code in this PR, I think the whole else/if can be removed (and thus deduplicating code), by putting the following in front of the `if (value instanceof Scriptable && ....`:
```
if (value instanceof BigInteger) {
   value = ScriptRuntime.toObject(state.cx, state.scope, value);
}
```

```

Reply:
You are wrong here, the stuble difference is not to change value in case of bigint but no toJSON. Otherwise the replacer will be called with a NativeBigInt instead of a java BigInt (and this leads to failing tests).
Will have to check why NativeBigint does not overwrite getTypeof() - but this is a different story.

Reply:
Right